File: data_labeling_examples/bulk_labeling_java/src/main/java/com/oracle/datalabelingservicesamples/scripts/RemoveLabelScript.java
package com.oracle.datalabelingservicesamples.scripts;


import com.oracle.bmc.datalabelingservicedataplane.model.Annotation;
import com.oracle.bmc.datalabelingservicedataplane.model.AnnotationSummary;
import com.oracle.bmc.datalabelingservicedataplane.model.Dataset;
import com.oracle.bmc.datalabelingservicedataplane.model.Entity;
import com.oracle.bmc.datalabelingservicedataplane.model.GenericEntity;
import com.oracle.bmc.datalabelingservicedataplane.model.Label;
import com.oracle.bmc.datalabelingservicedataplane.model.LabelName;
import com.oracle.bmc.datalabelingservicedataplane.model.UpdateAnnotationDetails;
import com.oracle.bmc.datalabelingservicedataplane.requests.DeleteAnnotationRequest;
import com.oracle.bmc.datalabelingservicedataplane.requests.GetAnnotationRequest;
import com.oracle.bmc.datalabelingservicedataplane.requests.GetDatasetRequest;
import com.oracle.bmc.datalabelingservicedataplane.requests.ListAnnotationsRequest;
import com.oracle.bmc.datalabelingservicedataplane.requests.UpdateAnnotationRequest;
import com.oracle.bmc.datalabelingservicedataplane.responses.GetAnnotationResponse;
import com.oracle.bmc.datalabelingservicedataplane.responses.GetDatasetResponse;
import com.oracle.bmc.datalabelingservicedataplane.responses.ListAnnotationsResponse;
import com.oracle.datalabelingservicesamples.requests.Config;
import com.oracle.datalabelingservicesamples.requests.RemoveLabel;
import lombok.extern.slf4j.Slf4j;

import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

@Slf4j
public class RemoveLabelScript extends RemoveLabel {

    static Dataset dataset;
    static String removeLabelPrefix;
    static Set<String> actualDatasetLabels = new HashSet<>();
    static Set<String> labelsToRemove = new HashSet<>();
    static ExecutorService executorService;
    static AtomicInteger successCount;
    static AtomicInteger failureCount;

    static {
        executorService = Executors.newFixedThreadPool(Config.INSTANCE.getThreadCount());
        successCount = new AtomicInteger();
        failureCount = new AtomicInteger();
    }

    public static void main(String [] args) throws ExecutionException, InterruptedException {
        log.info("...Starting to remove labels...");
        long startTime = System.nanoTime();
        String datasetId = Config.INSTANCE.getDatasetId();
        validateRequest(datasetId);
        removeLabelPrefix = Config.INSTANCE.getRemoveLabelPrefix();

        if(removeLabelPrefix.equals("*")){
            //delete all annotations
            for (String label : actualDatasetLabels){
                    labelsToRemove.add(label);
            }
        }
        else{
            for (String label : actualDatasetLabels){
                if(label.startsWith(removeLabelPrefix)){
                    labelsToRemove.add(label);
                }
            }
            if(labelsToRemove.isEmpty()){
                log.error("No Label found with prefix {}",removeLabelPrefix );
                throw new InvalidParameterException("Invalid Input RemoveLabelPrefix Provided");
            }
        }
        log.info("Labels to be removed");
        log.info(String.valueOf(labelsToRemove));
        String page = null;

        do{
            ListAnnotationsRequest listAnnotationsRequest = ListAnnotationsRequest.builder().compartmentId(dataset.getCompartmentId()).datasetId(datasetId).lifecycleState(Annotation.LifecycleState.Active).page(page).build();
            ListAnnotationsResponse response = Config.INSTANCE.getDlsDpClient().listAnnotations(listAnnotationsRequest);
            assert response.get__httpStatusCode__() == 200 : "List Annotation Response Fail";

            if(response.getAnnotationCollection().getItems().size()>0){
                List<CompletableFuture<Void>> completableFutures = new ArrayList<>();
                for(AnnotationSummary annotation : response.getAnnotationCollection().getItems()){
                    CompletableFuture<Void> future = CompletableFuture
                            .runAsync(() -> processAnnotationSummaryAndRemoveLabel(annotation),executorService);
                    completableFutures.add(future);
                }
                CompletableFuture<Void> combinedFuture = CompletableFuture
                        .allOf(completableFutures.toArray(new CompletableFuture[0]));
                combinedFuture.get();
            }
            page = response.getOpcNextPage();
        }while(page != null);
        executorService.shutdown();
        log.info("Total annotations to process : {}, "
                        +"Successfully processed {}, "
                        +"Failed to process {}",
                successCount.get() + failureCount.get(),
                successCount.get(),
                failureCount.get());
        long elapsedTime = System.nanoTime() - startTime;
        log.info("Time Taken is {} second(s)", elapsedTime / 1_000_000_000);
    }

    private static void validateRequest(String datasetId) {
        /*
         * Validate Dataset
         */
        GetDatasetRequest getDatasetRequest = GetDatasetRequest.builder().datasetId(datasetId).build();
        GetDatasetResponse datasetResponse = Config.INSTANCE.getDlsDpClient().getDataset(getDatasetRequest);
        assert datasetResponse.get__httpStatusCode__() == 200 : "Invalid Dataset Id Provided";
        List<LabelName> datasetLabelSet = datasetResponse.getDataset().getLabelSet().getItems();

        log.info("Dataset Labels : ");
        for (LabelName labelName : datasetLabelSet) {
            actualDatasetLabels.add(labelName.getName());
            log.info(labelName.getName());
        }
        dataset = datasetResponse.getDataset();
    }
    private static void processAnnotationSummaryAndRemoveLabel(AnnotationSummary annotationSummary){
        GetAnnotationRequest getAnnotationRequest = GetAnnotationRequest.builder().annotationId(annotationSummary.getId()).build();
        GetAnnotationResponse response = Config.INSTANCE.getDlsDpClient().getAnnotation(getAnnotationRequest);
        assert response.get__httpStatusCode__() == 200 : "GetAnnotation Request Failed";

        List<Entity> entityList = response.getAnnotation().getEntities();
        if (entityList.size() > 1) {
            log.error("Single/Multi label annotation can have only one annotation entity");
            return;
        }

        for (Entity entity : entityList) {
            if (entity instanceof GenericEntity) {
                GenericEntity genericEntity = (GenericEntity) entity;
                List<Label> entityLabels = genericEntity.getLabels();
                List<Label> newLabelList = new ArrayList<>();
                for(Label label : entityLabels){
                    if(!labelsToRemove.contains(label.getLabel())){
                        newLabelList.add(label);
                    }
                }
                if(newLabelList.isEmpty()){
                    deleteAnnotation(response.getAnnotation().getId());
                }
                else if(newLabelList.equals(entityLabels)){
                    log.info("Nothing to update");
                }
                else{
                    GenericEntity newEntity = GenericEntity.builder().labels(newLabelList).extendedMetadata(genericEntity.getExtendedMetadata()).build();
                    updateAnnotation(response.getAnnotation(),newEntity);
                }
            }
        }
    }
    private static void updateAnnotation(Annotation annotation, Entity entity){
        List<Entity> entityList = new ArrayList<>();
        entityList.add(entity);
        try {
            UpdateAnnotationDetails updateAnnotationDetails = UpdateAnnotationDetails.builder().definedTags(annotation.getDefinedTags()).freeformTags(annotation.getFreeformTags()).entities(entityList).build();
            UpdateAnnotationRequest updateAnnotationRequest = UpdateAnnotationRequest.builder().annotationId(annotation.getId()).updateAnnotationDetails(updateAnnotationDetails).build();
            Config.INSTANCE.getDlsDpClient().updateAnnotation(updateAnnotationRequest);
            successCount.incrementAndGet();
        }catch(Exception e){
            log.error("Exception in update annotation");
            failureCount.incrementAndGet();

        }
    }
    private static void deleteAnnotation(String annotationId){
        try {
            DeleteAnnotationRequest deleteAnnotationRequest = DeleteAnnotationRequest.builder().annotationId(annotationId).build();
            Config.INSTANCE.getDlsDpClient().deleteAnnotation(deleteAnnotationRequest);
            successCount.incrementAndGet();
        } catch (Exception e){
            log.error("Exception in delete annotation");
            failureCount.incrementAndGet();
        }
    }
}

File: data_labeling_examples/bulk_labeling_python/remove_label_script.py
import oci
import os
from config import *
from oci.data_labeling_service_dataplane.data_labeling_client import DataLabelingClient
from oci.data_labeling_service_dataplane.models import GenericEntity, Entity, Annotation, AnnotationCollection, AnnotationSummary, UpdateAnnotationDetails
import time
import logging
import json
from concurrent.futures import ThreadPoolExecutor
from concurrent.futures import as_completed
from concurrent.futures import wait

def init_dls_dp_client(config, service_endpoint):
    dls_client = DataLabelingClient(config=config, service_endpoint=service_endpoint)
    return dls_client


config_file = oci.config.from_file(CONFIG_FILE_PATH, CONFIG_PROFILE)
retry_strategy = oci.retry.DEFAULT_RETRY_STRATEGY
dls_dp_client = init_dls_dp_client(config_file, SERVICE_ENDPOINT_DP)
labels_to_remove = []

def process_annotation_and_remove_label(logger, annotation_id):
    try:
        response = dls_dp_client.get_annotation(annotation_id=annotation_id)
    except Exception as error:
            global failure_count
            failure_count+=1
            response = error
    if response.status == 200:
        entities = response.data.entities

        if len(entities) > 1:
            print("Single/Multi label annotation can have only one annotation entity")
            return

        for e in entities:
            entity_type = e.entity_type
            extended_metadata = e.extended_metadata
            entityLabel = e.labels

        newLabelList = []

        for label in entityLabel:
            if not label.label in labels_to_remove:
                newLabelList.append(label)

        if not newLabelList:
            delete_annotation(annotation_id)
        elif newLabelList == entityLabel:
            print("Nothing to update")
        else :
            new_entity = [GenericEntity(entity_type=entity_type, extended_metadata=extended_metadata, labels=newLabelList)]
            update_annotation(response.data, new_entity)

def delete_annotation(annotation_id):
    try:
        response = dls_dp_client.delete_annotation(annotation_id=annotation_id)
        global success_count
        success_count+=1
    except Exception as error:
        print(error)
        global failure_count
        failure_count+=1

def update_annotation(annotation, entity):
    update_annotation_details_obj = UpdateAnnotationDetails(defined_tags=annotation.defined_tags ,entities=entity, freeform_tags=annotation.freeform_tags)
    try:
        response = dls_dp_client.update_annotation(annotation_id=annotation.id, update_annotation_details=update_annotation_details_obj)
        global success_count
        success_count+=1
    except Exception as error:
        print(error)
        global failure_count
        failure_count+=1

def main():

    logging.basicConfig(filename="debug.log",
                            format='%(asctime)s %(message)s',
                            filemode='w')

    logger = logging.getLogger()
    logger.setLevel(logging.DEBUG)
    logger.info("...........starting to remove labels.............")
    start = time.perf_counter()
    try:
        response = dls_dp_client.get_dataset(dataset_id=DATASET_ID)
        logger.info("Fetching Dataset")
    except Exception as error:
        response = error
    if response.status == 200:
        logger.info("Fetching Dataset Successful")
        compartment_id = response.data.compartment_id
        # manage the json
        data = json.loads(str(response.data.label_set))
        dataset_labels = [labels["name"] for labels in data["items"]]
        print("Dataset Labels : ")
        print(dataset_labels)
        for label in dataset_labels:
            if label.startswith(REMOVE_LABEL_PRIFIX):
                labels_to_remove.append(label)

        if not labels_to_remove:
            print(f"No Label found with prefix " + REMOVE_LABEL_PRIFIX)
            return
        print("Labels to be removed : ")
        print(labels_to_remove)
        global success_count
        global failure_count
        success_count=0
        failure_count=0
        page = None

        while True:
            annotation_response = dls_dp_client.list_annotations(compartment_id=compartment_id, dataset_id=DATASET_ID, lifecycle_state="ACTIVE",page=page)
            annotation_data = json.loads(str(annotation_response.data))
            annotation_ids = [annotation["id"] for annotation in annotation_data["items"]]

            with ThreadPoolExecutor(NO_OF_PROCESSORS) as executor:
                futures = [executor.submit(process_annotation_and_remove_label, logger, annotation_id) for annotation_id in annotation_ids]
            wait(futures)
            if annotation_response.has_next_page :
                page = annotation_response.next_page
            else:
                break

    end = time.perf_counter()
    print(f'Total annotations to process : {(success_count + failure_count)}\n'
          f'Successfully processed {success_count}\n'
          f'Failed to process {failure_count}\n'
          f'Finished in {round(end - start, 2)} second(s)')

if __name__ == "__main__":
    main()